File: misc/spotify/official-solutions/greedycow_js.java
/* This is an example solution to the "Greedy Cows" problem from
 * Spotify Code Quest 2012, held on Feb 18, 2012.
 *
 * The solution is provided as-is, is not documented, and may be
 * generally difficult to read.
 *
 * This work by Scrool (http://scrool.se/) is licensed under a
 * Creative Commons Attribution-Share Alike 2.5 Sweden License
 * (http://creativecommons.org/licenses/by-sa/2.5/se/deed.en).  You
 * are free to redistribute it as you like as long as this license
 * information is not removed.
 */
import java.io.BufferedInputStream;
import java.io.InputStream;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class greedycow_js {

	private Scanner in;
	private PrintStream out;

	public greedycow_js(InputStream stream, PrintStream out) {
		in = new Scanner(new BufferedInputStream(stream));
		this.out = out;
	}

	long n;
	int f;
	List<Pair<Long, Long>> points = new ArrayList<Pair<Long, Long>>(f);

	private void indata() {
		n = in.nextLong();
		f = in.nextInt();
		for (int i = 0; i < f; i++) points.add(new Pair(in.nextLong(), in.nextLong()));
	}

	private void removeColinear() {
		for (int i = 0; i < (1000 * points.size()) && points.size() >= 3; i++) {
			Pair<Long, Long> a = points.get(i % points.size());
			Pair<Long, Long> b = points.get((i + 1) % points.size());
			Pair<Long, Long> c = points.get((i + 2) % points.size());
			//vector a -> b
			long x1 = b.first - a.first;
			long y1 = b.second - a.second;
			//vector b -> c
			long x2 = c.first - b.first;
			long y2 = c.second - b.second;

			//check if cross product is zero
			long cprod = x1 * y2 - y1 * x2;
			if (cprod == 0) {
				points.remove((i + 1) % points.size());
			}
		}
		f = points.size();
	}

	long[][] binom = binomialTable(100000, 5);

	public boolean ok(int ans) {
		int points1 = ans / f;
		int count1 = f - (ans % f);
		int points2 = points1+1;
		int count2 = ans % f;
		int totPoints = points1 * count1 + points2 * count2;
		assert totPoints == ans;
		long points = binom[totPoints][4] + binom[totPoints][2] + 1;
		points -= ((binom[points1][2] + binom[points1][3] * (totPoints - points1) + binom[points1][4])) * count1;
		points -= ((binom[points2][2] + binom[points2][3] * (totPoints - points2) + binom[points2][4])) * count2;
		return points >= n;
	}

	private void solve() {
		indata();
		removeColinear();
		int ans = 0;
		while (!ok(ans)) ans++;
		out.println(ans);
	}

	public static long[][] binomialTable(int n, int k) {
		long[][] binoms = new long[n + 1][k + 1];
		for (int i = 0; i <= n; i++) {
			binoms[i][0] = 1;
			for (int j = 1; i > 0 && j <= k; j++) binoms[i][j] = (binoms[i - 1][j - 1] + binoms[i - 1][j]);
		}
		return binoms;
	}

	public static void main(String... args) {
		new greedycow_js(System.in, System.out).solve();
	}

	public class Pair<A, B> {

		public A first;
		public B second;

		public Pair(A first, B second) {
			this.first = first;
			this.second = second;
		}
	}


}

File: misc/spotify/official-solutions/greedycow_js.cpp
/* This is an example solution to the "Greedy Cows" problem from
 * Spotify Code Quest 2012, held on Feb 18, 2012.
 *
 * The solution is provided as-is, is not documented, and may be
 * generally difficult to read.
 *
 * This work by Scrool (http://scrool.se/) is licensed under a
 * Creative Commons Attribution-Share Alike 2.5 Sweden License
 * (http://creativecommons.org/licenses/by-sa/2.5/se/deed.en).  You
 * are free to redistribute it as you like as long as this license
 * information is not removed.
 */
#include <cstdio>
#include <vector>

using namespace std;

typedef long long int ll;
typedef pair<ll, ll> post;

ll n, F;
vector<post> Fs;

ll bindp[100000][5];

ll bin(int n, int k) {
	return bindp[n][k];
}

ll c(int points1, int count1, int points2, int count2){
	int pp = points1*count1 + points2*count2;
	ll points = bin(count1*points1 + count2*points2, 4) + bin(count1*points1 + count2*points2, 2) + 1;
	points -= (bin(points1, 2) + bin(points1, 3)*(pp-points1) +bin(points1, 4)) * count1;
	points -= (bin(points2, 2) + bin(points2, 3)*(pp-points2) +bin(points2, 4)) * count2;
	return points;
}

int main(){
	for(int nb = 0; nb<100000; nb++){
		for(int kb = 0; kb<=4; kb++){
			if(kb == 0) bindp[nb][kb] = 1;
			else if (nb == 0) bindp[nb][kb] = 0;
			else bindp[nb][kb] = bindp[nb-1][kb-1] + bindp[nb-1][kb];
		}
	}
	scanf("%lld%lld", &n, &F);
	for(int i = 0; i<F; i++){
		ll x, y;
		scanf("%lld%lld", &x, &y);
		Fs.push_back(post(x, y));
	}
	for(size_t i = 1; i<=Fs.size()*3; i++){
		post prev = Fs[(i-1+Fs.size())%Fs.size()];
		post cur = Fs[(i+Fs.size())%Fs.size()];
		post next = Fs[(i+1+Fs.size())%Fs.size()];
		ll dx1 = cur.first - prev.first;
		ll dx2 = next.first - cur.first;
		ll dy1 = cur.second - prev.second;
		ll dy2 = next.second - cur.second;
		if(dx1*dy2 == dy1*dx2){
			Fs.erase(Fs.begin()+i);
			i--;
		}
	}
	F = Fs.size();
	int minpoints = 0;
	int hasmin = F;
	while(c(minpoints, hasmin, minpoints+1, F-hasmin) < n){
		hasmin--;
		if(hasmin == 0){
			minpoints++;
			hasmin = F;
		}
	}
	printf("%lld\n", minpoints*hasmin + (minpoints+1)*(F-hasmin));
}
